core/src/main/scala/org/bitcoins/core/crypto/MnemonicCode.scala

```diff
@@ -0,0 +1,257 @@
+
+      // tail-recursive binary search for word in word list
+      // stolen from https://codereview.stackexchange.com/questions/161887/re-inventing-the-wheel-binary-search-in-scala
+      import MnemonicCode.{ENGLISH_WORDS => wordlist}
+      @tailrec
+      def search(
+          start: Int = 0,
+          end: Int = wordlist.length - 1
+      ): Option[Int] = {
+        val mid = start + (end - start) / 2
+        if (start > end) None // can't be found
+        else if (wordlist(mid) == word) Some(mid) // found
+        else if (wordlist(mid) > word)
+          search(start, mid - 1) // narrow the field
+        else search(mid + 1, end)
+      }
+
+      val indexOpt = search()
+
+      val index = indexOpt.getOrElse(
+        throw new IllegalArgumentException(s"Couldn't find $word in word list!")
+      )
+
+      BitVector.fromInt(index, size = MnemonicCode.BIT_GROUP_LENGTH)
+
+    }
+
+    val bits: BitVector = words
+      .map(curr => bitsForWord(curr))
+      .fold(BitVector.empty)((accum, curr) => accum ++ curr)
+
+    val codeInfo = MnemonicCode.getMnemonicCodeInfo(words)
+
+    bits.take(codeInfo.entropyAndChecksumBits)
+  }
+}
+
+/**
+  * @see [[https://github.com/bitcoin/bips/blob/master/bip-0039.mediawiki BIP39]]
+  */
+object MnemonicCode {
+
+  private[crypto] val BIT_GROUP_LENGTH = 11
+
+  private case class MnemonicCodeImpl(words: Vector[String])
+      extends MnemonicCode
+
+  /**
+    * Creates a mnemonic code from the provided words
+    */
+  def fromWords(words: Vector[String]): MnemonicCode = MnemonicCodeImpl(words)
+
+  /**
+    * Generates a mnemonic code from the given entropy bytes
+    */
+  def fromEntropy(entropy: ByteVector): MnemonicCode =
+    fromEntropy(entropy.bits)
+
+  /**
+    * The minimum number of entropy bits needed to construct a mnemonic code
+    */
+  private[crypto] val MIN_ENTROPY_BITS = 128
+
+  /**
+    * The maximum number of entropy bits allowed to construct mnemonic code
+    */
+  private[crypto] val MAX_ENTROPY_BITS = 256
+
+  /**
+    * The entropy for a mnemonic code has to be a multiple of 32
+    */
+  private[crypto] val ENTROPY_MULTIPLE = 32
+
+  /**
+    * Generates a mnemonic code from the given entropy bits
+    */
+  def fromEntropy(entropy: BitVector): MnemonicCode = {
+    require(
+      entropy.length >= MIN_ENTROPY_BITS,
+      s"Entropy must be at least $MIN_ENTROPY_BITS bits"
+    )
+
+    require(
+      entropy.length <= MAX_ENTROPY_BITS,
+      s"Entropy cannot be longer than $MAX_ENTROPY_BITS bits"
+    )
+
+    require(
+      entropy.length % ENTROPY_MULTIPLE == 0,
+      s"Entropy must be a multiple of $ENTROPY_MULTIPLE bits!"
+    )
+
+    val ENTROPY_CHECKSUM_DIVISOR = 32
+    val checksumLength = entropy.length / ENTROPY_CHECKSUM_DIVISOR
+
+    val hashedEntropy = CryptoUtil.sha256(entropy)
+    val hashedEntropyBits = hashedEntropy.bytes.toBitVector
+    val checkSum = hashedEntropyBits.take(checksumLength)
```

Comment:
Shouldn't we be validating the checksum here?

Reply:
The checksum isn't provided by the user, it's calculated by us and then used to get the final word in our phrase. Then later on (at the top of the abstract class) , we validate it.